### PR TITLE
[fei4413.2.usingfetchpolicies] Implement fetch policy in useCachedEffect

### PR DESCRIPTION
**Pull request by somewhatabstract**
## Summary:
This reworks `useCachedEffect` to add fetch policy and refetch support.

This was a heavier lift than I originally hoped for, however, it works. I think things will get a little nicer when the RequestFulfillment properly supports aborting requests, but that's a while away (see FEI-4276) as we need to determine how to exactly do that. For now, I've added basic "aborting" to the RequestFulfillment that we'll have to expand later.

Issue: FEI-4413

## Test plan:
`yarn test`

### CODE REVIEW

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 1]
```diff
@@ -81,8 +96,9 @@ export const useCachedEffect = <TData: ValidCacheData>(
     options: CachedEffectOptions<TData> = ({}: $Shape<
         CachedEffectOptions<TData>,
     >),
-): Result<TData> => {
+): [Result<TData>, () => void] => {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:99]: Now returning the result and a `refetch` method in a tuple.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 2]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
-
-            // If our caller provided a cacheUpdated callback, we use that.
-            // Otherwise, we toggle our little state update.
-            if (onResultChanged != null) {
-                onResultChanged(value);
-            } else {
-                forceUpdate();
+    // For the NetworkOnly fetch policy, we ignore the cached value.
+    // So we need somewhere else to store the network value.
+    const networkResultRef = React.useRef();
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:126]: For the NetworkOnly fetch policy, we need a way to track that network result outside of the shared cache. I did try using state and replacing the force update but I was getting some infinite render loop issues and I couldn't quite track it down. So for now, I'll do it this way where the value is stored as a ref and we just look at it on render to determine what to do.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 3]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+
+    // Set up the function that will do the fetching.
+    const currentRequestRef = React.useRef();
+    const fetchRequest = React.useMemo(() => {
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:130]: We move this out of the `useEffect` so we can reference it for our `refetch` ability.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 4]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+        // We aren't using useCallback here because we need to make sure that
+        // if we are rememo-izing, we cancel any inflight request for the old
+        // callback.
+        currentRequestRef.current?.cancel();
+        currentRequestRef.current = null;
+        networkResultRef.current = null;
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:136]: We use `useMemo` so that when we're asked to recreate the callback, we have a chance to reset some stuff from any previous invocation.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 5]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+            const request = RequestFulfillment.Default.fulfill(requestId, {
+                handler: interceptedHandler,
+            });
+
+            if (request === currentRequestRef.current?.request) {
+                // The request inflight is the same, so do nothing.
+                // NOTE: Perhaps if invoked via a refetch, we will want to
+                // override this behavior and force a new request?
+                return;
+            }
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:157]: If we get invoked more than once (like someone calling `refetch` twice in a row or something), this just protects us from adding our `.then` handler more than once.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 6]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+
+            // Clear the last network result.
+            networkResultRef.current = null;
+
+            // Cancel the previous request.
+            currentRequestRef.current?.cancel();
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:163]: So, a `refetch` occurs but the inflight request is different. I don't "think" this is technically possible since the callback itself would've been recreated, but being defensive to make sure.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 7]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+
+            // TODO(somewhatabstract, FEI-4276):
+            // Until our RequestFulfillment API supports cancelling/aborting, we
+            // will have to do it.
+            let cancel = false;
```
Comment (somewhatabstract) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:168]: When aborts are real aborts that give AbortError responses, I don't believe we'll need to guard the `then` with a cancel token anymore. That'll be nice. Once day.

**Review by jeresig — APPROVED** [review]
Looks good to me! Glad we are building up parity on this.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 8]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+                    cancel = true;
+                    RequestFulfillment.Default.abort(requestId);
+                },
+            };
+        };
+        // We deliberately ignore the handler here because we want folks to use
+        // interceptor functions inline in props for simplicity. This is OK
+        // since changing the handler without changing the requestId doesn't
+        // really make sense - the same requestId should be handled the same as
+        // each other.
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:211]: This feels like it could admit subtle issues if the interceptor function captures variables from its closure.  Is this the existing behavior?

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js (hunk 4 above)
Comment (kevinbarabash) [reply packages/wonder-blocks-data/src/hooks/use-cached-effect.js:136]: Thanks for adding this comment about the use of `useMemo` over `useCallback`.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 9]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+                );
             }
-        },
-        [setMostRecentResult, onResultChanged, forceUpdate],
-    );
+            // We use our request fulfillment here so that in-flight
+            // requests are shared.
+            const request = RequestFulfillment.Default.fulfill(requestId, {
+                handler: interceptedHandler,
+            });
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:150]: Is `fulfill` memoized?

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 10]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+    const fetchRequest = React.useMemo(() => {
+        // We aren't using useCallback here because we need to make sure that
+        // if we are rememo-izing, we cancel any inflight request for the old
+        // callback.
+        currentRequestRef.current?.cancel();
+        currentRequestRef.current = null;
+        networkResultRef.current = null;
+
+        // Now we can return the new fetch function.
+        return () => {
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:139]: style nit: It took me a while to find this `return` statement.  Assigning this to a variable and returning a variable from the bottom of the `useMemo()` might improve the readability a bit.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 11]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+        // interceptor functions inline in props for simplicity. This is OK
+        // since changing the handler without changing the requestId doesn't
+        // really make sense - the same requestId should be handled the same as
+        // each other.
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [
+        requestId,
+        onResultChanged,
+        forceUpdate,
+        setMostRecentResult,
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:217]: I know that setters from `useState()` don't need to be listed.  I wonder if `setMostRecentResult` falls into that category.  The lint rule isn't smart enough to know unfortunately.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 12]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+        fetchPolicy,
+    ]);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:219]: This feels like it could be extracted into a separate `useFetch()` hook.  Not necessary for this PR, but maybe something for a followup PR.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 13]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
-            // least once and so we already made the request at least once. So
-            // we can bail out right here.
-            return true;
+    // indicates its a different request.
+    const requestIdRef = React.useRef(requestId);
+
+    // Calculate if we want to fetch the result or not.
+    // If this is true, we will do a new fetch, cancelling the previous fetch
+    // if there is one inflight.
+    const shouldFetch = React.useMemo(() => {
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:228]: I like how this logic has been consolidated.  It reads quite nicely.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js [hunk 14]
```diff
@@ -104,107 +120,152 @@ export const useCachedEffect = <TData: ValidCacheData>(
...
+    }, [requestId, mostRecentResult, fetchPolicy, hardSkip]);
 
-        return false;
-    }, [requestId, previousRequestId, mostRecentResult]);
+    // Let's make sure our ref is set to the most recent requestId.
+    requestIdRef.current = requestId;
 
-    // So now we make sure the client-side request happens per our various
-    // options.
     React.useEffect(() => {
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/use-cached-effect.js:259]: The effect is comically simple now, love it.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js [hunk 15]
```diff
@@ -91,49 +100,94 @@ describe("#useCachedEffect", () => {
             },
         );
 
-        it("should not request data", () => {
-            // Arrange
-            const fakeHandler = jest.fn().mockResolvedValue("data");
+        it.each(Array.from(FetchPolicy.members()))(
+            "should not request data for FetchPolicy.%s",
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:104]: Could you add something to the title about the reason for it not requesting data?  It's because `fakeHandler` is resolved with data, right?

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js [hunk 16]
```diff
@@ -190,13 +229,137 @@ describe("#useCachedEffect", () => {
...
+                } = clientRenderHook(() =>
+                    useCachedEffect("ID", fakeHandler, {fetchPolicy}),
+                );
+                fakeHandler.mockClear();
+                await act((): Promise<mixed> => response);
+                act(refetch);
+                await act((): Promise<mixed> => response);
+
+                // Assert
+                expect(fakeHandler).toHaveBeenCalledTimes(1);
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:257]: What's the connection between fakeHandler being called once and the test title?

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js [hunk 17]
```diff
@@ -190,13 +229,137 @@ describe("#useCachedEffect", () => {
...
+            // Assert
+            expect(underTest).toThrowErrorMatchingInlineSnapshot(
+                `"Cannot fetch with CacheOnly policy"`,
+            );
+        });
+
+        it.each(
+            Array.from(FetchPolicy.members()).filter(
+                (e) => e !== FetchPolicy.CacheOnly,
+            ),
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:286]: style nit: assigning this value to `AllFetchPoliciesExceptCacheOnly` would help with readability or list them out in an array like you do in the next `it.each()`.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js [hunk 18]
```diff
@@ -449,59 +616,158 @@ describe("#useCachedEffect", () => {
...
+                let renderCount = 0;
+                const Component = React.memo(() => {
+                    const [, refetch] = useCachedEffect("ID", fakeHandler, {
+                        fetchPolicy,
+                    });
+                    React.useEffect(() => {
+                        refetch();
+                        refetch();
+                        refetch();
+                        refetch();
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:667]: These get collapsed to a single request right?

File: packages/wonder-blocks-data/src/hooks/__tests__/use-hydratable-effect.test.js [hunk 19]
```diff
@@ -407,7 +413,7 @@ describe("#useHydratableEffect", () => {
             rerender({requestId: "ID2"});
 
             // Assert
-            expect(result.current).toStrictEqual(Status.loading());
+            expect(result.current).toStrictEqual(Status.success("GOODDATA"));
```
Comment (kevinbarabash) [inline packages/wonder-blocks-data/src/hooks/__tests__/use-hydratable-effect.test.js:416]: Why did this behavior change?

**Review by kevinbarabash — APPROVED** [review]
The implementation changes look reasonable to me.  The tests are very comprehensive, but I do have a couple of questions about some of the changes in behavior in a couple of the tests.

I'm also a little concerned about the move to return an array instead of an object like before.  If you haven't done so already, you may want to do a pulse check with the 321 team about that detail publishing.

File: packages/wonder-blocks-data/src/__docs__/exports.use-cached-effect.stories.mdx [hunk 20]
```diff
@@ -16,20 +16,22 @@ function useCachedEffect<TData: ValidCacheData>(
     requestId: string,
     handler: () => Promise<TData>,
     options?: CachedEffectOptions<TData>,
-): Result<TData>;
+): [Result<TData>, () => void];
```
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/__docs__/exports.use-cached-effect.stories.mdx:19]: I don't think `Result` should have an API like that. It's a cacheable piece of state information. The hooks pattern is well-defined to have APIs that return methods in a tuple like this.

[the remark this answers is not in the document]
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/__docs__/exports.use-cached-effect.stories.mdx:19]: Also, `Result` is used in many more places than just this hook - it wouldn't make sense to have a `refetch` in those other circumstances.

File: packages/wonder-blocks-data/src/__docs__/exports.use-cached-effect.stories.mdx [hunk 21]
```diff
@@ -16,20 +16,22 @@ function useCachedEffect<TData: ValidCacheData>(
...
 ```
 
-This hook is a wrapper around `useEffect`. Internally it uses the [`useSharedCache`](/docs/data-exports-usesharedcache--page) hook to store the result of the effect and, if that result is available, returns it without rerunning the effect. The precise behavior of the
-hook can be modified using the options.
+This hook invokes the given handler and caches the result using the [`useSharedCache`](/docs/data-exports-usesharedcache--page) hook. The `requestId` is used to both share identify inflight requests that can be shared, and to identify the cached value to use.
 
-There is currently no way to reinvoke the effect without changing the `requestId`.
+The hook returns an array containing the current state of the requested data, and a function that can be used to `refetch` the data on demand. Note that calling `refetch` while an inflight request is in progress for the given `requestId` will be a no-op.
```
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/__docs__/exports.use-cached-effect.stories.mdx:24]: I disagree. THat would be like the `useState` method returning the current state with a `set` function on it. It doesn't make sense. The current result is different from the API to update the current result.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js (hunk 15 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:104]: No, it's because it's server-side. That's the entire context here. This hook should never invoke the effect when server-side. I feel like adding that to the test name duplicates information already held in the parent describe.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js (hunk 16 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:257]: Just being explicit. When refetch is called, the handler should be called. Seems fair to also expect it to only invoke the handler once in that scenario.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js (hunk 18 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/__tests__/use-cached-effect.test.js:667]: Exactly. That's what the test is asserting.

File: packages/wonder-blocks-data/src/hooks/__tests__/use-hydratable-effect.test.js (hunk 19 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/__tests__/use-hydratable-effect.test.js:416]: Because the previous test case was actually copypasta and not doing what the test title stated. I updated it to have cached data for the new request ID to assert that it does indeed do what the test title says.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js (hunk 9 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/use-cached-effect.js:150]: No, but why would it need to be? It's not being recreated inline or anything. It's just defined function.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js (hunk 8 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/use-cached-effect.js:211]: Yes, it is.

Since interception is only used in fixtures and tests, I don't think that subtlety is an issue and we can call it out in docs if it becomes one.

File: packages/wonder-blocks-data/src/hooks/use-cached-effect.js (hunk 11 above)
Comment (somewhatabstract) [reply packages/wonder-blocks-data/src/hooks/use-cached-effect.js:217]: Yeah, I don't know, but those setters are memoized, so it doesn't seem to be an issue. :)
